### PR TITLE
Vectorized a bunch of the discrete rngs

### PR DESCRIPTION
**Pull request by bbbales2**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Vectorized a bunch of the discrete random number generators

As far as vectorizing all the rngs goes, this skips all the distributions that don't have vectorized pmfs/pdfs or don't have rngs. ~~There are a few distributions where only the interface is being tested in the vectorized tests (no distributions). This includes von_mises, pareto type 2, and exp_mod_normal.~~

Edit: This should have the rest of the vectorized rngs for https://github.com/stan-dev/math/issues/45. Anything missing is unintentional.

#### Intended Effect:

Same as above

#### How to Verify:

```
./runTests.py test/unit/math/prim/mat/prob
./runTests.py test/unit/math/prim/scal/prob
```

#### Side Effects:

Hopefully none!

#### Documentation:

In the doxygen

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): University of California, Santa Barbara

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by roualdes** [comment]
@bbbales2, this is great.  Can't wait for it to be merged.  When it is, I'll expose these on the Stan side.

**Comment by bbbales2** [comment]
I had a look at the multivariate stuff. The only three distributions that actually had vectorized lpdfs were multi_normal, multi_normal_cholesky, and multi_student_t, so I just manually built tests for those.

Had to add a few new metaprograms to make this work: StdVectorBuilder, contains_std_vector, and is_std_vector. The difference from VectorBuilder, contains_vector, and is_vector is that the only thing StdVectorBuilder respects as a vectorized type are actual std::vectors. Dunno if there is a better way of doing this.

I also added some rudimentary distribution tests to von_mises, pareto type 2, and exp_mod_normal (which didn't previously have tests).

At this point this pull req. should have all the stuff for https://github.com/stan-dev/math/issues/45 . Any missing tests/vectorizations are a bug.

**Comment by bob-carpenter** [comment]
This is fantastic.  Let me know when it's ready to review.

**Comment by bob-carpenter** [comment]
Great.  I'm going to wait until the Jenkins issues get ironed out.  We just rebooted the server.  So I'll try retesting.

Jenkins, retest this, please.

**Comment by bbbales2** [comment]
@mitzimorris has kindly agreed to review this pull req.

Here's the User's Guide to the pull:

The bulk of this is implementation and testing of vectorized integer rngs, but there are also a few multivariate distributions and updates to the continuous ones.

A prototypical continuous univariate rng is the normal distribution
Implementation: https://github.com/stan-dev/math/blob/develop/stan/math/prim/scal/prob/normal_rng.hpp
Tests: https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/test/unit/math/prim/mat/prob/normal_test.cpp

A prototypical discrete univariate rng is the poisson distribution
Implementation: https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/stan/math/prim/scal/prob/poisson_rng.hpp
Tests: https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/test/unit/math/prim/mat/prob/poisson_test.cpp

A prototypical multivariate distribution is multi_normal_cholesky:
Implementation: https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/stan/math/prim/mat/prob/multi_normal_cholesky_rng.hpp
Tests: https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/test/unit/math/prim/mat/prob/multi_normal_cholesky_test.cpp

Implementation of the different _rng interfaces is simplified through the use of:
scalar_seq_view, vector_seq_view, VectorBuilder, and StdVectorBuilder. StdVectorBuilder is new to this pull req. and is used by the multivariate rngs to construct the correct output (https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/stan/math/prim/scal/meta/StdVectorBuilder.hpp). There are two additional metaprograms, contains_std_vector and is_std_vector which are new. They're modeled after contains_vector and is_vector.

Testing for the univariate rngs works by building test rigs and then feeding them to two of three types of test functions which all live in https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
check_dist_throws_* - which tests the interface works and the code throws errors in the expected places
check_quantiles_* - which tests that continuous rngs are sampling from the expected distributions
check_counts_* - which tests that discrete rngs are sampling from the expeced dists.

Each rng makes a test rig that inherits from either https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/test/unit/math/prim/mat/prob/VectorRealRNGTestRig.hpp (for reals) or https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/test/unit/math/prim/mat/prob/VectorIntRNGTestRig.hpp forints. They inherit from https://github.com/bbbales2/math/blob/0c9ef5559181a7598bd199f50446c48778824e9e/test/unit/math/prim/mat/prob/VectorRNGTestRig.hpp.

**Comment by bbbales2** [comment]
Following up to @mitzimorris 

I hunted down the finicky quantile (https://github.com/stan-dev/math/pull/622#issuecomment-334510950). I think it's just that, and so I reported a bug to Boost https://svn.boost.org/trac10/ticket/13620.

File: test/unit/math/prim/mat/meta/contains_std_vector_test.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  EXPECT_TRUE(contains_std_vector<std::vector<Eigen::VectorXd> >::value);
+  EXPECT_TRUE(contains_std_vector<std::vector<Eigen::RowVectorXd> >::value);
+  bool r = contains_std_vector<std::vector<
+      Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> > >::value;
+  EXPECT_TRUE(r);
+}
+
+TEST(MetaTraits, contains_vector_false) {
+  using stan::contains_std_vector;
+
```
[the remark this answers is not in the document]
Comment (bbbales2) [reply test/unit/math/prim/mat/meta/contains_std_vector_test.cpp]: Those are in test/unit/math/prim/arr/meta/contains_std_vector_test.cpp . Hard to keep all these folders in order :P.

File: stan/math/prim/scal/meta/StdVectorBuilder.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ *  for constructing the output
+ *
+ *  The difference between this and VectorBuilder is that VectorBuilder
+ *  counts Eigen::Matrix<T, Dynamic, 1> and Eigen::Matrix<T, 1, Dynamic>
+ *  types as vectors.
+ *
+ *  @tparam used boolean variable indicating whether this instance
+ *      is used. If this is false, there is no storage allocated
+ *      and operator[] throws.
+ *  @tparam T1 Type of vector to build
```
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/scal/meta/StdVectorBuilder.hpp]: I updated the docs! This should be good to go.

**Comment by bbbales2** [comment]
Phew lol, finally got this to pass, after updating a comment. @mitzimorris want to merge?

**Comment by mitzimorris** [comment]
merged!

**Comment by roualdes** [comment]
Awesome.  I'd like to get to the Stan side of this by the end of this week.  Thanks to you both.

**Comment by bbbales2** [comment]
Cool! Just pm'ed you on Discourse to ask if you were still up for it haha. Thanks for looking at this!

**Pull request merged**